# Preppin' Data
## 2024: Week 5 - Getting the right data
**Created by:** Jenny Martin | [Challenge Link](https://preppindata.blogspot.com/2024/01/2024-week-5-getting-right-data.html)

Prep Air are interested in creating a workflow that has multiple outputs depending on user requirements. They want users to be able to answer the following questions:<br>
<br>
What are the details of the customers who have booked flights and which routes are they travelling on?<br>
Which customers are yet to book a flight in 2024?<br>
Which flights are yet to be booked by customers in 2024?<br>
The datasets you'll be working with are fairly large so you'll need to decide which tables to join (and when) to be as efficient as possible.<br>

In [1]:
# Input the data
# There are 3 tables to connect to for this challenge

import pandas as pd
import datetime as dt
flights = pd.read_csv("Prep Air 2024 Flights.csv", parse_dates=["Date"], date_format="%d/%m/%Y").sort_index()
flights

Date Flight Number      From        To
0     2024-11-22         PA001    London  New York
1     2024-11-23         PA001    London  New York
2     2024-11-23         PA002  New York    London
3     2024-11-24         PA001    London  New York
4     2024-11-27         PA001    London  New York
...          ...           ...       ...       ...
2509  2024-11-19         PA007  New York     Perth
2510  2024-11-20         PA007  New York     Perth
2511  2024-11-20         PA009  New York     Tokyo
2512  2024-11-20         PA010     Tokyo  New York
2513  2024-11-20         PA011     Perth     Tokyo

[2514 rows x 4 columns]

In [2]:
customers = pd.read_csv("Prep Air Customers.csv", parse_dates=["Last Date Flown"], date_format="%d/%m/%Y").sort_index()
customers

Customer ID Last Date Flown first_name  last_name  \
0               1      2023-01-05     Denyse     Gebuhr   
1               2      2023-10-05      Keene   Devennie   
2               3      2023-11-09      Tyler    McGrail   
3               4      2023-11-22      Drusi     Ibeson   
4               5      2023-12-23   Stanwood   Seacroft   
...           ...             ...        ...        ...   
9994         9995      2024-01-19    Hesther  Braidwood   
9995         9996      2023-09-16     Jelene   Dodgshun   
9996         9997      2023-11-26        Ira       Duff   
9997         9998      2023-03-12    Yalonda  Carrivick   
9998         9999      2023-06-29      Netta     Course   

                          email  gender  
0          dgebuhr0@vinaora.com  Female  
1        kdevennie1@plala.or.jp    Male  
2             tmcgrail2@nyu.edu    Male  
3        dibeson3@hostgator.com  Female  
4     sseacroft4@wikispaces.com    Male  
...                         ...     ...  
9994   hbraidwoodrm@reuters.com  Female  
9995  jdodgshunrn@angelfire.com  Female  
9996      iduffro@delicious.com    Male  
9997   ycarrivickrp@samsung.com  Female  
9998       ncourserq@weebly.com  Female  

[9999 rows x 6 columns]

In [3]:
sales = pd.read_csv("Prep Air Ticket Sales.csv", parse_dates=["Date"], date_format="%d/%m/%Y").sort_index()
sales

Date Flight Number  Customer ID  Ticket Price
0      2024-01-03         PA001          232        818.99
1      2024-01-03         PA001          293       1947.99
2      2024-01-03         PA001          472       1350.99
3      2024-01-03         PA001          572        905.99
4      2024-01-03         PA001         1191        567.99
...           ...           ...          ...           ...
44763  2024-11-17         PA012         5949       1404.99
44764  2024-11-17         PA012         9547        838.99
44765  2024-11-18         PA011         9109       1930.99
44766  2024-11-18         PA012         9109        187.99
44767  2024-11-20         PA011         2814       1182.99

[44768 rows x 4 columns]

### For the first output:
Create a dataset that gives all the customer details for booked flights in 2024. Make sure the output also includes details on the flights origin and destination<br>
When outputting the data, create an excel file with a new sheet for each output (so 1 file for all outputs this week!)

In [4]:
# joining the tables with sales as the left table

combined = sales.merge(flights, how="left", on=["Date","Flight Number"]).merge(customers, how="left", on="Customer ID")
combined

Date Flight Number  Customer ID  Ticket Price    From        To  \
0      2024-01-03         PA001          232        818.99  London  New York   
1      2024-01-03         PA001          293       1947.99  London  New York   
2      2024-01-03         PA001          472       1350.99  London  New York   
3      2024-01-03         PA001          572        905.99  London  New York   
4      2024-01-03         PA001         1191        567.99  London  New York   
...           ...           ...          ...           ...     ...       ...   
44763  2024-11-17         PA012         5949       1404.99   Tokyo     Perth   
44764  2024-11-17         PA012         9547        838.99   Tokyo     Perth   
44765  2024-11-18         PA011         9109       1930.99   Perth     Tokyo   
44766  2024-11-18         PA012         9109        187.99   Tokyo     Perth   
44767  2024-11-20         PA011         2814       1182.99   Perth     Tokyo   

      Last Date Flown first_name  last_name                       email  \
0          2024-01-03        Arv  Ballinger  aballinger6f@hostgator.com   
1          2024-01-03     Corney      Arger            carger84@mac.com   
2          2024-01-03       Katy      Akram    kakramd3@marketwatch.com   
3          2024-01-03      Jocko    Bolwell     jbolwellfv@hubpages.com   
4          2024-01-03     Gennie       Head       ghead5a@imageshack.us   
...               ...        ...        ...                         ...   
44763      2023-08-12        Oby    Dannehl        odannehlqc@upenn.edu   
44764      2023-03-10      Gregg   Mumbeson      gmumbesonf6@rambler.ru   
44765      2023-11-10     Marian    Yukhtin       myukhtin30@amazon.com   
44766      2023-11-10     Marian    Yukhtin       myukhtin30@amazon.com   
44767      2024-01-09      Debee     Kenion        dkenionml@disqus.com   

           gender  
0            Male  
1            Male  
2          Female  
3            Male  
4          Female  
...           ...  
44763        Male  
44764  Non-binary  
44765      Female  
44766      Female  
44767      Female  

[44768 rows x 11 columns]

### For the second output:
Create a dataset that allows Prep Air to identify which flights have not yet been booked in 2024<br>
Add a datestamp field to this dataset for today's date so that Prep Air know the unbooked flights as of the day the workflow is run<br>
When outputting the table to a new sheet in the Excel Workbook, choose the option "Append to Table" under Write Options. <br>
This means that if the workflow is run on a different day, the results will add additional rows to the dataset, rather than overwriting the previous run's data<br>

In [5]:
# joining the flights and sales table, with flights on outer left

flights_without_sales = flights.merge(sales, how="left", on=["Date", "Flight Number"], indicator=True)
flights_without_sales = flights_without_sales[flights_without_sales["_merge"] == "left_only"].drop(columns=["_merge", "Ticket Price", "Customer ID"])
flights_without_sales

Date Flight Number      From        To
0    2024-11-22         PA001    London  New York
1    2024-11-23         PA001    London  New York
2    2024-11-23         PA002  New York    London
3    2024-11-24         PA001    London  New York
4    2024-11-27         PA001    London  New York
..          ...           ...       ...       ...
291  2024-12-27         PA012     Tokyo     Perth
292  2024-12-28         PA012     Tokyo     Perth
293  2024-12-29         PA011     Perth     Tokyo
294  2024-12-30         PA011     Perth     Tokyo
295  2024-12-31         PA011     Perth     Tokyo

[296 rows x 4 columns]

In [7]:
# Adding a data stamp for today's date

flights_without_sales["Flights unbooked as of"] = pd.to_datetime("today").date()
flights_without_sales

Date Flight Number      From        To Flights unbooked as of
0    2024-11-22         PA001    London  New York             2025-10-10
1    2024-11-23         PA001    London  New York             2025-10-10
2    2024-11-23         PA002  New York    London             2025-10-10
3    2024-11-24         PA001    London  New York             2025-10-10
4    2024-11-27         PA001    London  New York             2025-10-10
..          ...           ...       ...       ...                    ...
291  2024-12-27         PA012     Tokyo     Perth             2025-10-10
292  2024-12-28         PA012     Tokyo     Perth             2025-10-10
293  2024-12-29         PA011     Perth     Tokyo             2025-10-10
294  2024-12-30         PA011     Perth     Tokyo             2025-10-10
295  2024-12-31         PA011     Perth     Tokyo             2025-10-10

[296 rows x 5 columns]

### For the third output:
Create a dataset that shows which customers have yet to book a flight with Prep Air in 2024<br>
Create a field which will allow Prep Air to see how many days it has been since the customer last flew (compared to 31/01/2024)<br>
#### Categorise customers into the following groups:<br>
Recent fliers - flown within the last 3 months<br>
Taking a break - 3-6 months since last flight<br>
Been away a while - 6-9 months since last flight<br>
Lapsed Customers - over 9 months since last flight<br>

In [8]:
# joining the customers and sales table, with customers on outer left

customers_not_flown = customers.merge(sales, how="left", on=["Customer ID"], indicator=True)
customers_not_flown = customers_not_flown[customers_not_flown["_merge"] == "left_only"].drop(columns=["_merge", "Ticket Price"])
customers_not_flown

Customer ID Last Date Flown first_name last_name  \
11               3      2023-11-09      Tyler   McGrail   
40               9      2023-06-05     Binnie   Jeckell   
56              14      2023-05-02      Ashil    Tetlow   
57              15      2023-05-18        Ayn  Bengtson   
100             24      2023-02-15      Grace   Piesing   
...            ...             ...        ...       ...   
45908         9972      2023-10-14        Any   Mongeot   
45917         9975      2023-03-04      Idell    Clyant   
45927         9978      2023-04-21      Spike   Thaller   
45989         9989      2023-11-06     Robbie   Grugerr   
46020         9996      2023-09-16     Jelene  Dodgshun   

                           email   gender Date Flight Number  
11             tmcgrail2@nyu.edu     Male  NaN           NaN  
40       bjeckell8@123-reg.co.uk   Female  NaN           NaN  
56        atetlowd@woothemes.com   Female  NaN           NaN  
57      abengtsone@bloomberg.com   Female  NaN           NaN  
100          gpiesingn@zdnet.com   Female  NaN           NaN  
...                          ...      ...  ...           ...  
45908     amongeotqz@twitter.com     Male  NaN           NaN  
45917    iclyantr2@hostgator.com   Female  NaN           NaN  
45927          sthallerr5@gov.uk  Agender  NaN           NaN  
45989     rgrugerrrg@plala.or.jp   Female  NaN           NaN  
46020  jdodgshunrn@angelfire.com   Female  NaN           NaN  

[1260 rows x 8 columns]

In [9]:
# create new column 'Days Since Last Flown' since 31/01/2024

customers_not_flown["Last Date Flown"] = pd.to_datetime(customers_not_flown["Last Date Flown"])
customers_not_flown["Days Since Last Flown"] = (pd.Timestamp("2024-01-31") - customers_not_flown["Last Date Flown"]).dt.days
customers_not_flown.head()

Customer ID Last Date Flown first_name last_name  \
11             3      2023-11-09      Tyler   McGrail   
40             9      2023-06-05     Binnie   Jeckell   
56            14      2023-05-02      Ashil    Tetlow   
57            15      2023-05-18        Ayn  Bengtson   
100           24      2023-02-15      Grace   Piesing   

                        email  gender Date Flight Number  \
11          tmcgrail2@nyu.edu    Male  NaN           NaN   
40    bjeckell8@123-reg.co.uk  Female  NaN           NaN   
56     atetlowd@woothemes.com  Female  NaN           NaN   
57   abengtsone@bloomberg.com  Female  NaN           NaN   
100       gpiesingn@zdnet.com  Female  NaN           NaN   

     Days Since Last Flown  
11                      83  
40                     240  
56                     274  
57                     258  
100                    350

In [10]:
# categorize customers

def customer_category(days):
    if days <= 90:
        return "Recent fliers"
    elif 91 <= days <= 180:
        return "Taking a break"
    elif 181 <= days <= 270:
        return "Been away a while"
    else:
        return "Lapsed Customers"

customers_not_flown["Customer Category"] = customers_not_flown["Days Since Last Flown"].apply(customer_category)
customers_not_flown = customers_not_flown.drop(columns=["Date", "Flight Number"])
customers_not_flown.head()

Customer ID Last Date Flown first_name last_name  \
11             3      2023-11-09      Tyler   McGrail   
40             9      2023-06-05     Binnie   Jeckell   
56            14      2023-05-02      Ashil    Tetlow   
57            15      2023-05-18        Ayn  Bengtson   
100           24      2023-02-15      Grace   Piesing   

                        email  gender  Days Since Last Flown  \
11          tmcgrail2@nyu.edu    Male                     83   
40    bjeckell8@123-reg.co.uk  Female                    240   
56     atetlowd@woothemes.com  Female                    274   
57   abengtsone@bloomberg.com  Female                    258   
100       gpiesingn@zdnet.com  Female                    350   

     Customer Category  
11       Recent fliers  
40   Been away a while  
56    Lapsed Customers  
57   Been away a while  
100   Lapsed Customers

## Output 1
### 2024 Booked Flights
44,768 rows (44,769 including headers)<br>
**11 fields:**
- Date
- From
- To
- Flight Number
- Customer ID
- Last Date Flown
- first_name
- last_name
- email
- gender
- Ticket Price

In [11]:
combined.head()

Date Flight Number  Customer ID  Ticket Price    From        To  \
0  2024-01-03         PA001          232        818.99  London  New York   
1  2024-01-03         PA001          293       1947.99  London  New York   
2  2024-01-03         PA001          472       1350.99  London  New York   
3  2024-01-03         PA001          572        905.99  London  New York   
4  2024-01-03         PA001         1191        567.99  London  New York   

  Last Date Flown first_name  last_name                       email  gender  
0      2024-01-03        Arv  Ballinger  aballinger6f@hostgator.com    Male  
1      2024-01-03     Corney      Arger            carger84@mac.com    Male  
2      2024-01-03       Katy      Akram    kakramd3@marketwatch.com  Female  
3      2024-01-03      Jocko    Bolwell     jbolwellfv@hubpages.com    Male  
4      2024-01-03     Gennie       Head       ghead5a@imageshack.us  Female

## Output 2
### Unbooked Flights
296 rows (297 including headers)<br>
**5 fields:**
- Flight unbooked as of
- Date
- Flight Number
- From
- To

In [12]:
flights_without_sales.head()

Date Flight Number      From        To Flights unbooked as of
0  2024-11-22         PA001    London  New York             2025-10-10
1  2024-11-23         PA001    London  New York             2025-10-10
2  2024-11-23         PA002  New York    London             2025-10-10
3  2024-11-24         PA001    London  New York             2025-10-10
4  2024-11-27         PA001    London  New York             2025-10-10

## Output 3
### Customers Yet to Book in 2024
1,260 rows (1,261 including headers)<br>
**8 fields:**
- Customer ID
- Customer Category
- Days Since Last Flown
- Last Date Flown
- first_name
- last_name
- email
- gender

In [13]:
customers_not_flown.head()

Customer ID Last Date Flown first_name last_name  \
11             3      2023-11-09      Tyler   McGrail   
40             9      2023-06-05     Binnie   Jeckell   
56            14      2023-05-02      Ashil    Tetlow   
57            15      2023-05-18        Ayn  Bengtson   
100           24      2023-02-15      Grace   Piesing   

                        email  gender  Days Since Last Flown  \
11          tmcgrail2@nyu.edu    Male                     83   
40    bjeckell8@123-reg.co.uk  Female                    240   
56     atetlowd@woothemes.com  Female                    274   
57   abengtsone@bloomberg.com  Female                    258   
100       gpiesingn@zdnet.com  Female                    350   

     Customer Category  
11       Recent fliers  
40   Been away a while  
56    Lapsed Customers  
57   Been away a while  
100   Lapsed Customers

In [15]:
# Output 1 and 2: run only once, to generate the file with the non-appending sheets

# with pd.ExcelWriter("output-202405.xlsx") as excel_file:
#   combined.to_excel(excel_file, sheet_name="2024 Booked Flights", index=False)
#   customers_not_flown.to_excel(excel_file, sheet_name="Customers Yet to Book in 2024", index=False)

In [16]:
# Output 3: appends every time the output gets executed

with pd.ExcelWriter("output-202405.xlsx", mode='a', if_sheet_exists='replace') as excel_append:
    flights_without_sales.to_excel(excel_append, sheet_name="Unbooked Flights", index=False)